src/library/scala/Option.scala

```diff
@@ -434,7 +434,7 @@ sealed abstract class Option[+A] extends IterableOnce[A] with Product with Seria
    *  @see flatMap
    */
   @inline final def foreach[U](f: A => U): Unit = {
-    if (!isEmpty) f(this.get)
+    if (!isEmpty) { f(this.get); () }
```

Comment:
I don't remember the default behavior on the other PR, but `if b then f()` is inherently side-effecting. The other flag for switching behavior on `if` is if you don't like the default.

Sample real foreach pain in Map:

![image](https://user-images.githubusercontent.com/369425/181706510-affe251b-66f8-444c-a2ca-23471d9c08b6.png)

A better solution (as just occurred to me) would be annotation to mark the type `U` as `Uninteresting`.

Reply:
Lukas just tweaked an if branch that wound up changing an inferred type from Unit to Any. I'll review Dotty and maybe emulate under `-Xsource;3`, i.e., single-leg if is always Unit (instead of lub with synthetic unit else).